lib/munster/models/received_webhook.rb

```diff
@@ -1,23 +1,107 @@
+    # we need to provide access not only to the webhook payload (the HTTP request body, usually), but also
+    # to the rest of the HTTP request - such as headers and route params. For example, imagine you use
+    # a system where your multiple tenants (users) may receive webhooks from the same sender. However, you
+    # need to dispatch those webhooks to those particular tenants of your application. Instead of mounting
+    # Munster as an engine under a common route, you mount it like so:
+    #
+    #    post "/incoming-webhooks/:user_id/:service_id" => "munster/receive_webhooks#create"
+    #
+    # This way, the tenant ID (the `user_id`) parameter is not going to be provided to you inside the webhook
+    # payload, as the sender is not sending it to you at all. However, you do have that parameter in your
+    # route. When processing the webhook, it is important for you to know which tenant has received the
+    # webhook - so that you can manipulate their data, and not the data belonging to another tenant. With
+    # validation, it is important too - in such a multitenant setup every user is likely to have their own,
+    # specific signing secret that they have set up. To find that secret and compare the signature, you
+    # need access to that `user_id` parameter.
+    #
+    # To allow access to these, Munster allows the ActionDispatch::Request object to be persisted. The
+    # persistence is not 1:1 - the Request is a fairly complex object, with lots of things injected into it
+    # by the Rails stack. Not all of those injected properties (Rack headers) are marshalable, some of them
+    # depend on the Rails application configuration, etc. However, we do retain the most important things
+    # for webhooks to be correctly handled.
+    #
+    # * The HTTP request body
+    # * The headers set by the webserver and the downstream proxies
+    # * The request body and query string params, depending on the MIME type
+    # * The route params. These are set by Journey (the Rails router) and cannot be reconstructed from a "bare" request
+    #
+    # While this reconstruction is best-effort it might not be lossless. For example, there might be no access
+    # to Rack hijack, streaming APIs, the cookie jar or other more high-level Rails request access features.
+    # You will, however, have the basics in place - such as the params, the request body, the path params
+    # (as were decoded by your routes) etc. But it should be sufficient to do the basic tasks to process a webhook.
+    #
+    # @return [ActionDispatch::Request]
+    def request
+      raise MISSING_HEADERS_COLUMN_ERROR unless self.class.column_names.include?("request_headers")
+      raise MISSING_HEADERS_ERROR if request_headers.blank?
+      headers = try(:request_headers) || {}
+      ActionDispatch::Request.new(headers.merge!("rack.input" => StringIO.new(body.to_s.b)))
```

Comment:
**thought:** I'm kind of wondering:
- How big of a size all this data would take? In most cases, it will probably take up more space than body itself. Do we need to purge records from database in this case?
- Can this create issues with future version of rails with chages in ActionDispatch::Request? Why saving just headers is not enough?

Reply:
Re. size - as big as is necessary. We do need to strip the raw post body though, which I forgot to do - essentially we store the request body twice.

I am using `ActionDispatch::Request` for a number of reasons.

* It was already used in `valid?` in all handlers, and all our handlers which validate expect that API to be available on the passed argument
* `ActionDispatch::Request` provides access to `params`...
* ...and more importantly - to the path params, which get set by the Rails router.

While we did implement the original Munster handlers with the expectation that they will be re-parsing the request body themselves, keeping the entire request available allows Rails to take over for parsing, and folks can use `params` like they are used to in Rails controllers, which is also a nice UX improvement IMO.

None of these affordances are possible to realise by other means. The API for `ActionDispatch::Request.new` is public API in Rails - if it changes, it would be just the same change as other changes in Rails internals, and our tests should catch that it changed.

Reply:
> We do need to strip the raw post body though, which I forgot to do - essentially we store the request body twice.

In this case, it seems reasonable to keep body in `ActionDispatch::Request` and not store the body in a separate column.